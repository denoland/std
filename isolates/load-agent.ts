import { assert, posix } from '@utils'
import { AGENT_RUNNERS, IA, Triad } from '@/constants.ts'
import { type Agent } from '@/constants.ts'
import matter from 'gray-matter'

export const api = {
  load: {
    description:
      'load an agent from a relative filepath. The file must be a markdown file, ending in ".md"',
    type: 'object',
    additionalProperties: false,
    required: ['path'],
    properties: {
      path: {
        description: 'the relative filepath of the agent',
        type: 'string',
        pattern: '^(\\.\\/|[^\\/].*\\/)*[^\\/]+\\.md$',
      },
    },
  },
  loadAll: {
    description: 'load all the agents from a directory',
    type: 'object',
    additionalProperties: false,
    required: ['dir'],
    properties: {
      dir: {
        description: 'the relative directory path',
        type: 'string',
        pattern: '^(\\.\\/|[^\\/][^\\/]*)$',
      },
    },
  },
}
export interface Api {
  load: (params: { path: string }) => Promise<Agent>
  loadAll: (
    params: { dir: string },
  ) => Promise<{ path: string; agent: Agent }[]>
}

export const functions = {
  load: async ({ path }: { path: string }, api: IA) => {
    assert(path.endsWith('.md'), 'path must end with .md')
    const string = await api.read(path)
    const { data, content } = matter(string.trim())
    assert(typeof content === 'string', 'content missing')
    const defaults = {
      runner: AGENT_RUNNERS.CHAT,
      instructions: '',
    }

    const { pid, commit } = api
    const source: Triad = { path, pid, commit }
    const instructions = content.trim()
    const name = posix.basename(path, posix.extname(path))
    const loaded: Agent = { ...defaults, ...data, name, instructions, source }

    assertAgent(loaded)
    return loaded
  },
  loadAll: async ({ dir }: { dir: string }, api: IA) => {
    // TODO provide globs
    const agents: { name: string; agent: Agent }[] = []
    const files = await api.ls(dir)
    for (const file of files) {
      if (file.endsWith('.md')) {
        const path = posix.basename(file, posix.extname(file))
        const agent = await functions.load({ path }, api)
        agents.push({ name: agent.name, agent })
      }
    }
    return agents
  },
}

const assertAgent = (agent: Agent) => {
  if (!(typeof agent.instructions === 'string')) {
    throw new Error('instructions must be an array')
  }
  const { runner } = agent
  if (runner !== AGENT_RUNNERS.CHAT) {
    throw new Error('runner must be chat')
  }
  if (agent.description && typeof agent.description !== 'string') {
    throw new Error('description must be a string')
  }
  if (agent.config) {
    const { config } = agent
    if (
      config.model && config.model !== 'gpt-3.5-turbo' &&
      config.model !== 'gpt-4-turbo' && config.model !== 'gpt-4o' &&
      config.model !== 'gpt-4o-mini'
    ) {
      throw new Error('invalid model: ' + config.model)
    }
    if (
      agent.config.temperature &&
      (agent.config.temperature < 0 || agent.config.temperature > 1)
    ) {
      throw new Error('temperature must be between 0 and 1')
    }
  }
  if (agent.commands && !Array.isArray(agent.commands)) {
    throw new Error('commands must be an array')
  }
}
