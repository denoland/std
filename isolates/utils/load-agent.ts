import merge from 'lodash.merge'
import { assert, posix } from '@utils'
import { AGENT_RUNNERS, agentSchema, IA, Triad } from '@/constants.ts'
import { type Agent } from '@/constants.ts'
import matter from 'gray-matter'
import { loadTools } from '@/isolates/utils/ai-load-tools.ts'

export const loadString = async (path: string, string: string, api: IA) => {
  const name = posix.basename(path, posix.extname(path))
  const { pid, commit } = api
  const source: Triad = { path, pid, commit }

  const { data, content } = matter(string.trim())
  assert(typeof content === 'string', 'content missing')
  let instructions
  try {
    instructions = await expandLinks(content, api)
  } catch (error) {
    throw new Error('Error expanding links: ' + path + '\n' + error.message)
  }

  const base: Agent = {
    name,
    source,
    runner: AGENT_RUNNERS.CHAT,
    instructions,
    config: {
      model: 'gpt-4o-mini',
      tool_choice: 'auto',
      parallel_tool_calls: true,
    },
    commands: [],
    napps: [],
  }

  merge(base, data)
  const { success, data: agent, error } = o1Checker.safeParse(base)
  if (!success) {
    throw new Error('Error parsing agent: ' + path + '\n' + error.message)
  }

  const toolsPromise = loadTools(agent, api)

  try {
    await toolsPromise
    return agent
  } catch (error) {
    throw new Error('Error parsing agent: ' + path + '\n' + error.message)
  }
}

export const load = async (
  path: string,
  api: IA,
  overrides?: Partial<Agent>,
) => {
  assert(path.endsWith('.md'), 'path must end with .md')
  const string = await api.read(path)
  const agent = await loadString(path, string, api)
  return merge(agent, overrides)
}

export const loadAll = async (dir: string, api: IA) => {
  // TODO provide globs
  const agents: { path: string; agent: Agent }[] = []
  const files = await api.ls(dir)
  for (const file of files) {
    if (file.endsWith('.md')) {
      const path = posix.basename(file, posix.extname(file))
      const agent = await load(path, api)
      agents.push({ path, agent })
    }
  }
  return agents
}

const expandLinks = async (content: string, api: IA, path: string[] = []) => {
  content = content.trim()
  const links = extractLinks(content)
  const contents = await loadContent(links, api, path)
  const result = replaceLinksWithContent(content, links, contents)
  return result.trim()
}
const loadContent = async (links: string[], api: IA, path: string[]) => {
  const contents = new Map()
  for (const link of links) {
    if (path.includes(link)) {
      throw new Error('circular reference: ' + path.join(' -> '))
    }
    const content = await api.read(link)
    contents.set(link, await expandLinks(content, api, path.concat(link)))
  }
  return contents
}

const extractLinks = (content: string): string[] => {
  const linkRegex = /\[.*?\]\((.*?\.md)\)/g
  const plain = content.split(codeBlockRegex).filter((_, i) => i % 2 === 0)

  const links = []
  for (const section of plain) {
    let match
    while ((match = linkRegex.exec(section)) !== null) {
      links.push(match[1])
    }
  }
  return links
}

const codeBlockRegex = /(```[\s\S]*?```|`[^`\n]*`)/g
const replaceLinksWithContent = (
  content: string,
  links: string[],
  contents: Map<string, string>,
) => {
  const parts = content.split(codeBlockRegex)
  for (let i = 0; i < parts.length; i++) {
    if (i % 2 === 0) {
      for (const link of links) {
        const linkRegex = new RegExp(`\\[.*?\\]\\(${link}\\)`, 'g')
        const resolved = contents.get(link)
        assert(resolved, 'missing content for link: ' + link)
        parts[i] = parts[i].replace(linkRegex, resolved)
      }
    }
  }

  return parts.join('')
}

const o1Checker = agentSchema.refine((data) => {
  if (data.instructions) {
    if (['o1-preview', 'o1-mini'].includes(data.config.model)) {
      return !data.instructions
    }
  }
  return true
}, 'instructions are not allowed for o1 models')
