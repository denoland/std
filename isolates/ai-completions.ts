import { Debug } from '@utils'
import '@std/dotenv/load' // load .env variables
import OpenAI from 'openai'
import {
  Agent,
  AssistantMessage,
  ChatParams,
  CompletionMessage,
  Functions,
  getThreadPath,
  IA,
  print,
  Thread,
  toApi,
  ToApiType,
} from '@/constants.ts'
import { loadTools } from './ai-load-tools.ts'
import { load } from './utils/load-agent.ts'
import { z } from 'zod'

const log = Debug('AI:completions')

const apiKey = Deno.env.get('OPENAI_API_KEY')
if (!apiKey) {
  throw new Error('missing openai api key: OPENAI_API_KEY')
}
const ai = new OpenAI({ apiKey, timeout: 20 * 1000, maxRetries: 5 })

export const transcribe = async (file: File) => {
  const transcription = await ai.audio.transcriptions
    .create({
      file,
      model: 'whisper-1',
      prompt:
        'Backchat, GPT4, GPT3, Dreamcatcher, CRM, HAL, Deno, Stucks, Redlid',
    })
  return transcription.text
}

export const parameters = {
  complete: z.object({ path: z.string() }),
}
export const returns = {
  complete: z.void(),
}

export const api = toApi(parameters)

export type Api = ToApiType<typeof parameters, typeof returns>

export const functions: Functions<Api> = {
  async complete({ path }, api) {
    const threadPath = getThreadPath(api.pid)
    log('completing thread %o', threadPath, print(api.pid))
    const thread = await api.readThread(threadPath)

    const assistant = await complete(path, thread.messages, api)
    thread.messages.push(assistant)
    api.writeJSON(threadPath, thread)
    log('completion complete', assistant.tool_calls?.[0], assistant.content)
  },
}

const complete = async (
  path: string,
  messages: Thread['messages'],
  api: IA,
) => {
  const agent: Agent = await load(path, api)
  const tools = await loadTools(agent.commands, api)

  const args = getChatParams(agent, messages, tools)

  log('completion started with model: %o', args.model, print(api.pid))
  let retries = 0
  const RETRY_LIMIT = 5
  let errorMessage = ''
  while (retries++ < RETRY_LIMIT) {
    try {
      const completion = await ai.chat.completions.create(args)
      const result = completion.choices[0].message
      log('completion complete', agent.source.path, result)
      const assistant: AssistantMessage = {
        ...result,
        name: agent.source.path,
      }
      return assistant
    } catch (error) {
      console.error('ai completion error', error)
      errorMessage = error.message
    }
  }
  throw new Error(`Failed after ${retries} attempts: ${errorMessage}`)
}

export const getChatParams = (
  agent: Agent,
  messages: Thread['messages'],
  tools: OpenAI.ChatCompletionTool[],
) => {
  const { config } = agent
  const {
    model,
    temperature,
    tool_choice,
    parallel_tool_calls,
    presence_penalty,
  } = config
  const sysprompt: CompletionMessage = {
    role: 'system',
    content: agent.instructions + '\n\n' + additionInstructions(),
    name: agent.source.path,
  }
  const args: ChatParams = {
    model,
    temperature,
    messages: [...messages, sysprompt].map(safeAssistantName),
    seed: 1337,
    tools: tools.length ? tools : undefined,
    tool_choice: tools.length ? tool_choice : undefined,
    presence_penalty,
    parallel_tool_calls: tools.length ? parallel_tool_calls : undefined,
  }
  return args
}
const safeAssistantName = (message: CompletionMessage) => {
  if (message.role !== 'assistant' && message.role !== 'system') {
    return message
  }
  if (!message.name) {
    return message
  }
  if (!/^[a-zA-Z0-9_-]+$/.test(message.name)) {
    return { ...message, name: message.name.replaceAll(/[^a-zA-Z0-9_-]/g, '_') }
  }
  return message
}

const additionInstructions = () => {
  return 'The time is: ' + new Date().toISOString()
}
