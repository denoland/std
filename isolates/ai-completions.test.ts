import { loadAgent } from '@/isolates/utils/load-agent.ts'
import { complete } from '@/isolates/ai-completions.ts'
import { functions, safeAssistantName } from '@/isolates/ai-completions.ts'
import { expect, log } from '@utils'
import { createMockApi } from '@/tests/fixtures/mock-api.ts'

Deno.test('test the regex for agent name sanitization', () => {
  const result = safeAssistantName({
    role: 'assistant',
    name: 'agents/o1.md',
  })
  expect(result).toEqual({ role: 'assistant', name: 'agents_o1_md' })
})
Deno.test('generate images', async () => {
  const { api, stop } = await createMockApi('test/images')
  const path = 'images/test.png'
  const result = await functions.image({
    path,
    prompt: 'a dystopian robot overload mecha',
    lowQuality: true,
    size: '1024x1024',
    style: 'natural',
  }, api)
  const image = await api.readBinary(path)
  expect(image.byteLength).toBeGreaterThan(0)
  expect(image.byteLength).toEqual(result.size)
  stop()
})

Deno.test.only('inject empty', async (t) => {
  const { api, stop } = await createMockApi('test/inject-single')
  const path = 'fake/agent.md'
  api.write(path, '')
  const agent = await loadAgent(path, api)

  const { mock } = complete
  log.enable('AI:mocker')
  const actorId = 'test-actor-id'
  mock.useRecorder(actorId, t)

  const result = await complete(agent, [{
    role: 'user',
    content: 'say cheese in emoji',
    name: actorId,
  }], api)

  console.dir(result, { depth: 10 })

  stop()
})
