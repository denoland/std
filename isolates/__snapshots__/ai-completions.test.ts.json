{
  "inject empty": {
    "recordings": {
      "test-actor-id": [
        {
          "request": {
            "model": "gpt-4o-mini",
            "messages": [
              {
                "content": "say cheese in emoji",
                "role": "user",
                "name": "test-actor-id"
              }
            ],
            "seed": 1337
          },
          "reply": {
            "assistant": {
              "role": "assistant",
              "content": "Say cheese! ðŸ§€ðŸ˜„",
              "name": "fake/agent.md",
              "refusal": null
            },
            "stats": {
              "created": 1729038718,
              "model": "gpt-4o-mini-2024-07-18",
              "system_fingerprint": "fp_e2bde53e6e",
              "duration": 1270,
              "openAiProcessingMs": 950,
              "usage": {
                "completion_tokens": 8,
                "prompt_tokens": 15,
                "total_tokens": 23,
                "completion_tokens_details": {
                  "reasoning_tokens": 0
                },
                "prompt_tokens_details": {
                  "cached_tokens": 0
                }
              }
            }
          }
        }
      ]
    }
  }
}